src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java

```diff
@@ -588,85 +583,97 @@ public void close() {
+                  return;
+                }
+
+                Channel ch = chP.getNow();
+                ch.writeAndFlush(upload)
+                    .addListener(
+                        (f) ->{
+                          releaseUploadChannel(ch);
+                          if (f.isSuccess()) {
+                            result.set(null);
+                          } else {
+                            Throwable cause = f.cause();
+                            if (cause instanceof HttpException) {
+                              HttpResponse response = ((HttpException) cause).response();
+                              try {
+                                // If the error is due to an expired auth token and we can reset the input stream,
+                                // then try again.
+                                if (authTokenExpired(response) && reset(in)) {
+                                  refreshCredentials();
+                                  uploadAfterCredentialRefresh(upload, result);
+                                } else {
+                                  result.setException(cause);
+                                }
+                              } catch (IOException e) {
+                                result.setException(e);
+                              }
+                            } else {
+                              result.setException(cause);
+                            }
+                          }
+                        });
+              });
+      result.addListener(() -> Closeables.closeQuietly(in), MoreExecutors.directExecutor());
+      return result;
+    } else {
+      Closeables.closeQuietly(in);
+      return Futures.immediateFuture(null);
     }
   }
 
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void uploadAfterCredentialRefresh(UploadCommand upload, SettableFuture<Void> result) {
+    acquireUploadChannel()
+        .addListener(
+            (Future<Channel> chP) -> {
+              if (!chP.isSuccess()) {
+                result.setException(chP.cause());
+                return;
+              }
+
+              Channel ch = chP.getNow();
+              ch.writeAndFlush(upload)
+                  .addListener(
+                      (f) -> {
+                        releaseUploadChannel(ch);
+                        if (f.isSuccess()) {
+                          result.set(null);
+                        } else {
+                          result.setException(f.cause());
+                        }
+                      });
+            });
+  }
+
   @Override
   public ListenableFuture<Void> uploadFile(Digest digest, Path file) {
-    try (InputStream in = file.getInputStream()) {
-      uploadBlocking(digest.getHash(), digest.getSizeBytes(), in, /* casUpload= */ true);
-    } catch (IOException | InterruptedException e) {
+    try {
+      return uploadAsync(digest.getHash(), digest.getSizeBytes(), file.getInputStream(), true);
```

Comment:
nit: please re-add `/* casUpload= */`
(same a few lines below and line 695)

Reply:
Sorry about that. Not sure how that got lost.